### ISSUE TITLE
Test Che6 to Che7 upgrade using che-operator

### ISSUE THREAD

**Issue by skabashnyuk** [eclipse-che/che#14200]
### Is your task related to a problem? Please describe.
 Install Che 6.19 and then attempt to update it to 7.0 using an operator.
 Need to check
- Make sure  user db & workspace definitions in postgres, successfully upgrade the Che server to 7.0
- see if anything breaks... variables that got renamed? routes that are different? assumptions/APIs in Che 7 that changed?

**Comment by nickboldt** [comment]
Downstream: https://issues.jboss.org/browse/CRW-303

**Comment by metlos** [comment]
I tested the upgrade on both K8s (using minikube 1.3.1) and Openshift 3.11 (using minishift).

I've found the following minor problems:
1. Keycloak doesn't upgrade from `che-keycloak:6.19.0` to `che-keycloak:7.1.0` on operator upgrade. This seems to be caused by the explicitly mentioned Keycloak image version in the CR config.
1. When upgrading from che operator 0.5.0 (commit https://github.com/eclipse/che-operator/commit/32b8e159391d1882b12c20313c95fe151d730259) which is the last operator with a non-CRW version of Che6 to operator 7.1.0, the configuration of Keycloak looses the original DB password and tries to use a new generated one leading to failing upgrade, because Keycloak cannot start.

@davidfestal do you think those are problems worth creating issues for?

**Comment by skabashnyuk** [comment]
>those are problems worth creating issues for

CC @l0rd @davidfestal wdyt do we need to upgrade version of keycloak for already exists CheCluste installations?

**Comment by metlos** [comment]
> > those are problems worth creating issues for

> CC @l0rd @davidfestal wdyt do we need to upgrade version of keycloak for already exists CheCluste installations?

I've created https://github.com/eclipse/che/issues/14575 for that.

**Comment by davidfestal** [comment]
Sorry for the time before answering @metlos 

I would answer differently for these 2 cases:

Let me start with the second one:

> 2.When upgrading from che operator 0.5.0 (commit eclipse/che-operator@32b8e15) which is the last operator with a non-CRW version of Che6 to operator 7.1.0, the configuration of Keycloak looses the original DB password and tries to use a new generated one leading to failing upgrade, because Keycloak cannot start.

That seems a backward-compatibility bug to me. If there would be some real possibility (and compelling reason) that a user would still use this old che operator version, then we should look deeper into the root cause of this and possibly fix it.

> 1. Keycloak doesn't upgrade from che-keycloak:6.19.0 to che-keycloak:7.1.0 on operator upgrade. This seems to be caused by the explicitly mentioned Keycloak image version in the CR config.

This is an [already-known issue](https://github.com/eclipse/che/issues/13714) that was still there in previous versions of Che 7. This has been fixed in Che `7.1.0` for all upcoming upgrades (`7.1.0` to next releases) and also specifically for the upgrades from the `7.0.0` to `7.1.0`, since we knew that `7.0.0` was the last version with this bug (see PR https://github.com/eclipse/che-operator/pull/72#issue-310285011).

That's not possible to fix this for upgrades from **any** older version to `7.1.0` since users also might want to keep fixed versions into their `CheCluster` resources.

2 cases here:
- if we plan to use script for migration of CRW `1.2.x` operator to `2.0`, then we should be able to clean the frozen docker image versions in the `CheCluster` (set them back to empty string for Keycloak, Postgres, etc ...)  just after upgrading the operator to 7.1.0.
- if we plan to let OperatorHub + marketplace do the upgrade of the operator, then the user should have to **manually** clean the frozen docker image versions in the `CheCluster` (set them back to empty string for Keycloak, Postgres, etc ...)  just after upgrading the operator to 7.1.0.

@nickboldt @l0rd Does it seem an acceptable solution ?

If it is not acceptable, then we would need to implement the same type of automatic fix as we did for the 7.0.0 to 7.1.0 release:
- identify a single CRW 1.2.x release that is the one most CRW users would have (probably the last release one)=>  considered the same as the default one
- automatically clear the docker images in the `CheCluster`as soon as they correspond to the docker image of this single CRW release.

**Comment by metlos** [comment]
Discovered another issue - https://github.com/eclipse/che/issues/14630

**Comment by metlos** [comment]
After all, this affects 1 upgrade scenario:
1. Using Che6 based operator, install che cluster CR specifically mentioning the image tag of Che to install
1. Upgrade the operator to Che7 based one.
1. The Che server deployment doesn't change because the image tag is explicitly configured.
1. Now you need to manually change the tag to some 7.x version, which triggers #14630.

**Comment by davidfestal** [comment]
@nickboldt @l0rd ld you please provide your thoughts about the questions in comment https://github.com/eclipse/che/issues/14200#issuecomment-532384111 ?

This would be great to conclude on this and decide if any action item is required, before closing the analysis issue.

**Comment by davidfestal** [comment]
> After all, this affects 1 upgrade scenario:
> 
>     1. Using Che6 based operator, install che cluster CR specifically mentioning the image tag of Che to install
> 
>     2. Upgrade the operator to Che7 based one.
> 
>     3. The Che server deployment doesn't change because the image tag is explicitly configured.
> 
>     4. Now you need to manually change the tag to some 7.x version, which triggers #14630.

@metlos Do you ocnfirm that if you manually change the image tags in the custom resource by **patching** the CR existing in K8S instead of applying it again, upgrade works correctly ?

The main reason of my question is that, when you apply the changed CR again from the yaml file you have locally, you override the Postgres and Keycloak passwords to empty strings. This triggers the generation of new passwords by the operator, and brings bug #14630.

So to summarize, bug #14630 would only impact upgrade from 6 to 7 if the CR is changed fully, instead of only patching the docker image and tag fields.

**Comment by metlos** [comment]
> @metlos Do you confirm that if you manually change the image tags in the custom resource by patching the CR existing in K8S instead of applying it again, upgrade works correctly ?

Yes, patching the CR seems to be working well.

**Comment by l0rd** [comment]
@metlos @davidfestal the keycloak image version problem can be definitely handled as a doc issue (hence we need to add the workaround in the doc to close this issue).

**Comment by nickboldt** [comment]
> if we plan to use script for migration of CRW 1.2.x operator to 2.0, then we should be able to clean the frozen docker image versions in the CheCluster (set them back to empty string for Keycloak, Postgres, etc ...) just after upgrading the operator to 7.1.0.

I'm hoping to not have to handle scripting an update process, but if we have to we can do so as documentation, so it's not tightly tied to a release.

> if we plan to let OperatorHub + marketplace do the upgrade of the operator, then the user should have to manually clean the frozen docker image versions in the CheCluster (set them back to empty string for Keycloak, Postgres, etc ...) just after upgrading the operator to 7.1.0.

This sounds preferrable, except that we still have to support OCP 3.11 users. So would this mean we would tell them to install OLM in order to be able to use operator-based updates?

**Comment by davidfestal** [comment]
> 4. Now you need to manually change the tag to some 7.x version

@metlos Instead if patching the custom resource with some Che 7.x version, you should just reset the docker image to empty string, so that it takes the defaults.

Did you do this for all the following components:
- Keycloak (I saw you did)
- Postgres
- the PVC jobs docker image ?

I'm asking that because it would be important to have tested the upgrade of the Postgres image as well (if it needs upgrade).

**Comment by nickboldt** [comment]
+1, odds are good that with every CRW update we'll have new Keycloak, UBI8-minimal (PVC) and Postgres images because the hamster wheel of CVE fixes never stops. :)

**Comment by metlos** [comment]
I have not tested Postgres for sure because it hasn't changed between the versions for neither Che nor CRW. Not sure about PVC.

**Comment by skabashnyuk** [comment]
The scope of this task defined in the description is complete.
 Next steps 
- Explain pitfalls and workarounds of Che6 to Che7 upgrade in a doc https://github.com/eclipse/che/issues/14738
- Test UX of an upgrade to Che7 from the running Che6 workspace perspective  https://github.com/eclipse/che/issues/14810

I would like to ask no to extend the scope of this issue, instead of that create a new one.

**Comment by metlos** [comment]
> After che-operator report that Che is available again we have the following state:

I think we should test this:
1. Add custom tags for all the above mentioned images
1. Configure the CR to use those custom tags
1. After Che is redeployed with these new images
1. Set the image properties back to empty strings
1. The operator should use the default values for the images again.

It is basically what you have done with the difference that we're not creating new deployments but rolling over existing ones.

@davidfestal does the above make sense or do you think it is superfluous?

**Comment by nickboldt** [comment]
+1 to retesting with rolling updates.

Re: postgres updates...

> I have not tested Postgres for sure because it hasn't changed between the versions for neither Che nor CRW. Not sure about PVC.

Two problems with that statement:

a) [CRW 1.2.2](https://github.com/eclipse/che-operator/blob/1.2.2.GA/pkg/deploy/defaults.go#L35) is using `registry.redhat.io/rhscl/postgresql-96-rhel7:1-46` and 2.0 will be using `registry.redhat.io/rhscl/postgresql-96-rhel7:1-47` (or newer) so you CAN test a simple version change.

b) even with the upstream Che 6 -> 7 version `centos/postgresql-96-centos7:9.6` is a moving tag which represents the latest set of applied CVE fixes. So :9.6 from 2 months ago is a different version from today. You can see this from the tags in the repo which are 5 or 13 days old:

https://hub.docker.com/r/centos/postgresql-96-centos7/tags?page=1&ordering=last_updated

So... the flaw here is the approach that you're always using a moving tag, instead of a static one. 

Re: not sure about PVC, we can again test an update there:

* [1.2.2](https://github.com/eclipse/che-operator/blob/1.2.2.GA/pkg/deploy/defaults.go#L34
) / 6.19.x: registry.access.redhat.com/ubi8-minimal:8.0-159
vs.
* [2.0 / 7.3 / master](https://github.com/eclipse/che-operator/blob/master/pkg/deploy/defaults.go#L44-L45): registry.access.redhat.com/ubi8-minimal:8.0-213


So... reopening. Task not complete.

**Comment by nickboldt** [comment]
As requested followup work moved to 2 new issues: 

* https://github.com/eclipse/che/issues/14864 (Che 6.19.x to 7.x operator update)
* https://github.com/eclipse/che/issues/14865 (CRW 1.2 to 2.0 operator update)
